File: service/src/whf/ai/facts_tools.py
"""Read-only tools that let Copilot look up facts of one run. Pure functions first, Tool wrappers second."""

from __future__ import annotations

from typing import Any

from copilot import define_tool
from pydantic import BaseModel, Field

TOOL_NAMES = (
    "get_run_overview",
    "get_member_history",
    "get_member_forecast",
    "get_member_patterns",
    "get_member_open_tasks",
    "get_member_capacity",
    "get_project_timelines",
    "get_rebalancing_candidates",
)


class NoParams(BaseModel):
    """No parameters."""


class MemberParams(BaseModel):
    member_id: int = Field(description="The member id from get_run_overview")


class FactsToolbox:
    def __init__(self, facts: dict) -> None:
        self.facts = facts
        self._members = {int(m["id"]): m for m in facts.get("members", [])}

    # ----- pure lookups -------------------------------------------------
    def _unknown(self, member_id: int) -> dict:
        return {"error": f"unknown member_id {member_id}", "known_member_ids": sorted(self._members)}

    def run_overview(self) -> dict:
        f = self.facts
        return {
            "run": f["run"],
            "team": f["team"],
            "members": [{"id": m["id"], "name": m["name"], "role": m["role"]} for m in f["members"]],
            "model": f["model"],
            "rebalancing_candidates": f["rebalancing_candidates"],
            "how_to_proceed": "Call get_member_forecast, get_member_patterns, get_member_history, "
            "get_member_open_tasks and get_member_capacity for every member id listed here, "
            "then get_project_timelines, then answer with the JSON document.",
        }

    def member_history(self, member_id: int) -> dict:
        m = self._members.get(member_id)
        if m is None:
            return self._unknown(member_id)
        return {"member_id": member_id, "name": m["name"], "history_13w": m["history_13w"]}

    def member_forecast(self, member_id: int) -> dict:
        m = self._members.get(member_id)
        if m is None:
            return self._unknown(member_id)
        return {"member_id": member_id, "name": m["name"], "forecast": m["forecast"]}

    def member_patterns(self, member_id: int) -> dict:
        m = self._members.get(member_id)
        if m is None:
            return self._unknown(member_id)
        return {"member_id": member_id, "name": m["name"], "patterns": m["patterns"]}

    def member_open_tasks(self, member_id: int) -> dict:
        m = self._members.get(member_id)
        if m is None:
            return self._unknown(member_id)
        return {"member_id": member_id, "name": m["name"], "open_tasks": m["open_tasks"]}

    def member_capacity(self, member_id: int) -> dict:
        m = self._members.get(member_id)
        if m is None:
            return self._unknown(member_id)
        weeks = [
            {"week": row["week"], "capacity": row["capacity"], "demand": row["demand"], "overload": row["overload"]}
            for row in m["forecast"]
        ]
        return {"member_id": member_id, "name": m["name"], "weeks": weeks}

    def project_timelines(self) -> dict:
        return {"weeks": self.facts["run"]["weeks"], "projects": self.facts["projects"]}

    def rebalancing_candidates(self) -> dict:
        return dict(self.facts["rebalancing_candidates"])

    # ----- Copilot tools ------------------------------------------------
    def tools(self) -> list[Any]:
        box = self

        def member_tool(name: str, description: str, fn):
            return define_tool(
                name,
                description=description,
                handler=lambda params, _inv: fn(params.member_id),
                params_type=MemberParams,
                skip_permission=True,
            )

        def plain_tool(name: str, description: str, fn):
            return define_tool(
                name,
                description=description,
                handler=lambda _params, _inv: fn(),
                params_type=NoParams,
                skip_permission=True,
            )

        return [
            plain_tool(
                "get_run_overview",
                "Run, team, member list, model quality and rebalancing candidates. Call this first.",
                box.run_overview,
            ),
            member_tool(
                "get_member_history",
                "Last 13 weeks of task arrivals (hours and counts) for one member.",
                box.member_history,
            ),
            member_tool(
                "get_member_forecast",
                "Two-week forecast rows (demand, low, high, capacity, overload, open and new hours) for one member.",
                box.member_forecast,
            ),
            member_tool(
                "get_member_patterns",
                "Deterministic pattern statistics for one member (assignment style, weekday rhythm, trend, estimate bias, cycle time, lateness, cluster).",
                box.member_patterns,
            ),
            member_tool(
                "get_member_open_tasks",
                "Open tasks of one member with due dates and overdue flags.",
                box.member_open_tasks,
            ),
            member_tool(
                "get_member_capacity",
                "Capacity, demand and overload per forecast week for one member.",
                box.member_capacity,
            ),
            plain_tool(
                "get_project_timelines",
                "Projects of the team with start dates, deadlines and whether they start, end or run inside the forecast window.",
                box.project_timelines,
            ),
            plain_tool(
                "get_rebalancing_candidates",
                "Members with overload and members with spare capacity over the two weeks.",
                box.rebalancing_candidates,
            ),
        ]

File: server/forecast-core/src/main/java/com/workloadhub/forecast/ai/FactsTools.java
package com.workloadhub.forecast.ai;

import com.workloadhub.forecast.data.ExportFiles;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import tools.jackson.databind.JsonNode;

/** The nine tools Copilot may call, reading one run's stored facts and never the database. */
final class FactsTools {

    static final List<String> NAMES = List.of("get_run_overview", "get_member_history", "get_member_forecast", "get_member_patterns",
            "get_member_open_tasks", "get_member_capacity", "get_project_timelines", "get_rebalancing_candidates", "get_likely_work");

    private final JsonNode facts;
    private final Map<String, JsonNode> members = new TreeMap<>();

    FactsTools(JsonNode facts) {
        this.facts = facts;
        for (JsonNode m : facts.path("members")) {
            members.put(m.path("id").asText(), m);
        }
    }

    List<ToolSpec> specs() {
        return List.of(
                new ToolSpec("get_run_overview", "Run, team, member list, model quality, rebalancing candidates and data quality. Call this first.", false,
                        id -> runOverview()),
                new ToolSpec("get_member_history", "Last 13 weeks of task arrivals (hours and counts), logged hours of the last 4 weeks, unlogged and reopened tasks for one member.",
                        true, this::memberHistory),
                new ToolSpec("get_member_forecast", "Forecast rows per window (five weekdays each: demand, low, high, capacity, overload, due hours) and the day-by-day rows for one member.",
                        true, this::memberForecast),
                new ToolSpec("get_member_patterns", "Deterministic pattern statistics for one member (assignment style, weekday rhythm, trend, estimate bias, cycle time, lateness, cluster, backlog).",
                        true, this::memberPatterns),
                new ToolSpec("get_member_open_tasks", "Open tasks of one member with keys, estimates, due dates, overdue flags and project keys.", true,
                        this::memberOpenTasks),
                new ToolSpec("get_member_capacity", "Capacity, demand, overload and the two pressure figures (backlog_excess_hrs, due_excess_hrs)"
                        + " per forecast window and per day for one member, with working days and absence hours.", true,
                        this::memberCapacity),
                new ToolSpec("get_project_timelines", "Projects of the team with status, open and backlog task counts and the first due date, plus the forecast windows.",
                        false, id -> projectTimelines()),
                new ToolSpec("get_rebalancing_candidates", "The four lists of members to act on, over the run's windows: overloaded, underloaded"
                        + " (spare capacity), backlog_pressed (open work the run does not absorb) and deadline_pressed (work due inside a window"
                        + " beyond what the window holds).", false,
                        id -> rebalancingCandidates()),
                new ToolSpec("get_likely_work", "Per member, the likely-work signals: project roles held on live projects and the recent task-type mix.",
                        false, id -> likelyWork()));
    }

    private static Object plain(JsonNode node) {
        return node.isMissingNode() ? null : ExportFiles.mapper().treeToValue(node, Object.class);
    }

    private static Map<String, Object> map(Object... kv) {
        Map<String, Object> m = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) {
            m.put((String) kv[i], kv[i + 1]);
        }
        return m;
    }

    private Map<String, Object> unknown(String id) {
        return map("error", "unknown member_id " + id, "known_member_ids", new ArrayList<>(members.keySet()));
    }

    Map<String, Object> runOverview() {
        List<Object> list = new ArrayList<>();
        for (JsonNode m : facts.path("members")) {
            list.add(map("id", m.path("id").asText(), "name", m.path("name").asText(), "role", m.path("role").asText(null)));
        }
        @SuppressWarnings("unchecked")
        Map<String, Object> team = (Map<String, Object>) plain(facts.path("team"));
        return map("run", plain(facts.path("run")), "team", team, "members", list, "model", plain(facts.path("model")),
                "rebalancing_candidates", plain(facts.path("rebalancing_candidates")), "data_quality", plain(facts.path("data_quality")),
                "pending_holidays", plain(facts.path("pending_holidays")),
                "how_to_proceed", "Call get_member_forecast, get_member_capacity, get_member_patterns, get_member_history and get_member_open_tasks"
                        + " for every member id listed here, then get_project_timelines, get_likely_work and get_rebalancing_candidates, then answer with the JSON document.");
    }

    Map<String, Object> memberHistory(String id) {
        JsonNode m = members.get(id);
        if (m == null) {
            return unknown(id);
        }
        return map("member_id", id, "name", m.path("name").asText(), "history_13w", plain(m.path("history_13w")), "logged_hours_4w", plain(m.path("logged_hours_4w")),
                "unlogged_tasks", plain(m.path("unlogged_tasks")), "reopened_tasks", plain(m.path("reopened_tasks")));
    }

    Map<String, Object> memberForecast(String id) {
        JsonNode m = members.get(id);
        return m == null ? unknown(id)
                : map("member_id", id, "name", m.path("name").asText(), "forecast", plain(m.path("forecast")), "days", plain(m.path("days")));
    }

    Map<String, Object> memberPatterns(String id) {
        JsonNode m = members.get(id);
        return m == null ? unknown(id) : map("member_id", id, "name", m.path("name").asText(), "patterns", plain(m.path("patterns")));
    }

    Map<String, Object> memberOpenTasks(String id) {
        JsonNode m = members.get(id);
        return m == null ? unknown(id) : map("member_id", id, "name", m.path("name").asText(), "open_tasks", plain(m.path("open_tasks")));
    }

    Map<String, Object> memberCapacity(String id) {
        JsonNode m = members.get(id);
        if (m == null) {
            return unknown(id);
        }
        List<Object> windows = new ArrayList<>();
        for (JsonNode row : m.path("forecast")) {
            // The two pressure figures travel with overload (design 2026-09-13, section 8.2): this is the tool
            // that serves overload per window, and whf-forecast-interpretation tells Copilot to read the three
            // together — a zero overload beside an above-zero backlog_excess_hrs is not a member who is fine.
            windows.add(map("window", plain(row.path("window")), "start", row.path("start").asText(), "end", row.path("end").asText(),
                    "capacity", plain(row.path("capacity")), "demand", plain(row.path("demand")), "overload", plain(row.path("overload")),
                    "working_days", plain(row.path("working_days")), "absence_hours", plain(row.path("absence_hours")),
                    "backlog_excess_hrs", plain(row.path("backlog_excess_hrs")), "due_excess_hrs", plain(row.path("due_excess_hrs")),
                    "planned_hours", plain(row.path("planned_hours"))));
        }
        return map("member_id", id, "name", m.path("name").asText(), "windows", windows, "days", plain(m.path("days")));
    }

    Map<String, Object> projectTimelines() {
        return map("windows", plain(facts.path("run").path("windows")), "projects", plain(facts.path("projects")));
    }

    /**
     * The whole {@code rebalancing_candidates} node, as {@code get_run_overview} already serves it: whitelisting
     * keys here dropped {@code backlog_pressed} and {@code deadline_pressed} at the tool boundary (design
     * 2026-09-13, ruling 18.5), while whf-rebalancing-advice tells Copilot that {@code deadline_pressed} is the
     * strongest case for moving work.
     */
    @SuppressWarnings("unchecked")
    Map<String, Object> rebalancingCandidates() {
        Map<String, Object> node = (Map<String, Object>) plain(facts.path("rebalancing_candidates"));
        return node == null ? map() : node;
    }

    Map<String, Object> likelyWork() {
        List<Object> list = new ArrayList<>();
        for (JsonNode m : facts.path("members")) {
            JsonNode lw = m.path("likely_work");
            list.add(map("id", m.path("id").asText(), "name", m.path("name").asText(),
                    "project_roles", plain(lw.path("project_roles")), "recent_mix", plain(lw.path("recent_mix"))));
        }
        return map("members", list);
    }
}
